### PR TITLE
Fix #8874 - pdf works as expected now :)

### LINKED ISSUE #8874: Make pdf work :P
the good of pdf is that it works for linear functions, but if those contains basic blocks from different functions or so spaced ones it results in a very wrong output, where pdr is recommended.

The way to solve this is imho to add '...' in the parts of the disasm that doesnt belong to the function

### PR DESCRIPTION
**Pull request by radare**

### CODE REVIEW

File: libr/core/disasm.c [hunk 1]
```diff
@@ -4382,6 +4383,12 @@ static void ds_end_line_highlight(RDisasmState *ds) {
 	}
 }
 
+RAnalFunction *pdf = NULL;
```
Comment (kazarmy) [inline libr/core/disasm.c]: should be static

File: libr/core/cmd_print.c [hunk 2]
```diff
@@ -4293,8 +4293,15 @@ static int cmd_print(void *data, const char *input) {
...
 					} else {
-						r_core_cmdf (core, "pD %d @ 0x%08" PFMT64x,
-							f->_size > 0 ? f->_size: r_anal_fcn_realsize (f), f->addr);
+						ut64 at = f->addr;
+						ut64 sz = f->_size > 0 ? f->_size : r_anal_fcn_realsize (f);
+						ut8 *buf = calloc (sz, 1);
+						(void)r_io_read_at (core->io, at, buf, sz);
+						r_core_print_disasm_function (core, f);
+						core->num->value = r_core_print_disasm (core->print, core, at, buf, sz, sz, 0, 1, 0);
+						r_core_print_disasm_function (core, NULL);
```
Comment (kazarmy) [inline libr/core/cmd_print.c]: is the r_core_print_disasm_function() really needed? there are only a few places (okay, 6) where r_core_print_disasm() is called ... add an extra arg?

**Comment by radare** [comment]
Yeah this was kind of a quick poc. But i thinknadding more arguments is a bad idea. We should redesign this api.

I can add another arg and pass null to the others. But this func already have enough arguments :p

File: libr/core/cmd_print.c (hunk 2 above)
Comment (kazarmy) [reply libr/core/cmd_print.c]: this makes for more work if r_core_print_disasm() needs to be reentrant (if ever)

**Comment by radare** [comment]
Uh? Pdf is linear

**Comment by kazarmy** [comment]
> Yeah this was kind of a quick poc. But i thinknadding more arguments is a bad idea. We should redesign this api.
> 
> I can add another arg and pass null to the others. But this func already have enough arguments :p

ok 😁 

> Uh? Pdf is linear

I actually meant thread-safe but then I realized I had forgotten about thread-local variables.

File: libr/anal/fcn.c [hunk 3]
```diff
@@ -2132,7 +2132,7 @@ R_API RAnalBlock *r_anal_fcn_bbget(RAnalFunction *fcn, ut64 addr) {
 	RListIter *iter;
 	RAnalBlock *bb;
 	r_list_foreach (fcn->bbs, iter, bb) {
-		if (bb->addr == addr) {
+		if (addr >= bb->addr && addr < (bb->addr + bb->size)) {
```
Comment (ret2libc) [inline libr/anal/fcn.c]: this change is risky. This function was intended to return a RAnalBlock exactly at the given address.
For example, in disasm.c it is used to know whether an address was the start of a basicblock.

I suggest we create a separate function `r_anal_fcn_bbget_in`, like what we do for function where we have `_at` and `_in`.

**Comment by radare** [comment]
green <3

File: libr/anal/fcn.c (hunk 3 above)
Comment (ret2libc) [reply libr/anal/fcn.c]: yeah we could use a map to improve the performance, but until it's not a real problem i guess it's fine

**Pull request merged**